core/src/main/java/org/jboss/pnc/causeway/source/SourceRenamer.java

```diff
@@ -0,0 +1,99 @@
+        String gid = groupId.replace(".", "/");
+        String name = artifactId + "-" + version;
+        Path path = Paths.get(gid).resolve(artifactId).resolve(version);
+        return repack(input, name, path);
+    }
+
+    public RenamedSources repackNPM(InputStream input, String packageName, String version) throws CausewayException {
+        String name = packageName + "-" + version;
+        return repack(input, name, Paths.get(""));
+    }
+
+    private RenamedSources repack(InputStream input, String name, Path path) throws CausewayException {
+        try {
+            Path tempFile = Files.createTempFile("renamer-", ".tar.gz");
+
+            MessageDigest md5Digest = MessageDigest.getInstance("MD5");
+            DigestOutputStream outputStream = new DigestOutputStream(Files.newOutputStream(tempFile), md5Digest);
+            rewrite(input, outputStream, name);
+
+            BigInteger bi = new BigInteger(1, md5Digest.digest());
+            String md5Hash = String.format("%032x", bi);
+
+            String archiveName = name + ARCHIVE_SUFFIX;
+
+            return new RenamedSources(tempFile, path.resolve(archiveName).toString(), md5Hash);
+        } catch (IOException | CompressorException e) {
+            throw new CausewayException("Error while repacking archive with changed root directory name", e);
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("The JVM must support MD5 digest", e);
+        }
+    }
+
+    private void rewrite(InputStream input, OutputStream output, String name) throws CompressorException, IOException {
+        Path foo = Paths.get(name);
+        try (TarArchiveInputStream in = new TarArchiveInputStream(compressor.createCompressorInputStream(input));
+                TarArchiveOutputStream out = new TarArchiveOutputStream(new GzipCompressorOutputStream(output))) {
+            out.setLongFileMode(TarArchiveOutputStream.LONGFILE_GNU);
+
+            boolean rootFound = false;
+            for (TarArchiveEntry entry = in.getNextTarEntry(); entry != null; entry = in.getNextTarEntry()) {
+                Path originalName = Paths.get(entry.getName());
+                Path root = getTopmost(originalName);
+                if (root.equals(originalName)) { // validate only one directory exists in root of the archive
+                    if (!entry.isDirectory()) {
+                        throw new IllegalArgumentException("There is non-directory file in root of the archive.");
+                    }
+                    if (rootFound) {
+                        throw new IllegalArgumentException("Multiple directories in root of the archive.");
+                    }
+                    rootFound = true;
+                }
+                Path newName = foo.resolve(root.relativize(originalName));
+                entry.setName(newName.toString());
+                out.putArchiveEntry(entry);
+                IOUtils.copy(in, out);
```

Comment:
does this copy just the `entry` or entire `InputStream` to `OurputStream` ?

Reply:
It is really stupid. It copies the part of the original stream that represent the entry.

Reply:
that's extremely stupid... wtf